### PR TITLE
[GPU] Fix three cuDNN SDPA integration defects

### PR DESCRIPTION
**Pull request by YangXu1990uiuc**
# [GPU] Fix three cuDNN SDPA integration defects (use-after-free, missing output zero-fill under command buffers, determinism flag outside the cache key)

## Summary

Three defects in XLA's cuDNN fused-attention path, found while auditing cuDNN SDPA
consumers. The first is a use-after-free in the backward fMHA graph builder that
feeds a dangling `ScoreModFunc*` into the cuDNN frontend. The second means that
whenever an fMHA op is captured into a command buffer / CUDA graph, the output
zero-fill that XLA performs on the eager path is silently dropped, so the padded
tokens of `O` (and of `dQ`/`dK`/`dV`) return whatever the donated buffer happened
to contain. The third makes `force_deterministic` land in the backend config
*after* the instruction has been fingerprinted, so the compiled-graph lookup key
and the emitter's key can disagree under `--xla_gpu_deterministic_ops`.

### 1. Dangling `ScoreModFunc*` passed to the cuDNN frontend

`xla/backends/gpu/transforms/cudnn_custom_call_compiler.cc:361-377`
(`BuildGraphForCustomCallToBackwardFMHA`) built the score-mod functor inside an
`if` block and stored its address in a variable declared outside it:

```c++
    auto smf = stream_executor::gpu::ScoreModFunc(score_mod_fwd_comp,
                                                  score_mod_bwd_comp);
    score_mod = &smf;
```

`smf` is destroyed at the end of the `if` block, but `score_mod` is dereferenced
several statements later by
`se::gpu::GetCudnnFlashAttentionBackwardOperationGraph(..., score_mod)`, which
walks the two HLO computations to emit the score-modifier subgraph into the cuDNN
frontend graph. Every backward fMHA custom call that carries a score modifier
(any `jax.nn.dot_product_attention` / Flax attention lowered with a score-mod
computation) reads freed stack memory during compilation.

Fix: hoist the functor into a function-scope `std::optional` so it outlives the
graph build.

### 2. fMHA output zero-fill dropped on the command-buffer path

`xla/backends/gpu/runtime/cudnn_thunk.cc` — `CuDnnThunk::ExecuteOnStream` honours
`should_memzero_` and calls `stream->MemZero()` on every output buffer
(:106-108), but `CuDnnThunk::Record`, the command-buffer / CUDA-graph path, only
recorded the DNN graph launch. `should_memzero_` is set for exactly the fMHA
custom calls (`xla/service/gpu/thunk_emitter.cc:1240`,
`/*should_memzero=*/IsCustomCallTofMHA(*instr)`), so the flag exists precisely
because cuDNN does not write the whole buffer.

Why cuDNN behaves that way: the SDPA node applies a padding mask and honours the
`SEQ_LEN_Q` / `CU_SEQ_LEN_Q` / ragged-offset tensors
(`cudnn_frontend/include/cudnn_frontend/node/scaled_dot_product_flash_attention.h`,
the `padding_mask` / ragged-offset handling around :61-129 and :757-856). Tokens
outside `seq_len_q`, and any region past the ragged offsets, are simply never
stored — cuDNN's contract is that the consumer owns those bytes. XLA reuses
(donates) buffers aggressively, so those bytes are live garbage from a previous
op, not zeros.

Failure scenario: run a variable-length / padded SDPA under
`--xla_gpu_enable_command_buffer=FUSION,CUDNN` (or any config where the fMHA
thunk is converted into a command). Both `Record` branches — the explicit
`CreateDnnGraphCommand` path and the traced `RecordTracedCommand` fallback —
launch the graph with no preceding zero-fill, so the padded rows of `O` (fwd) and
of `dQ`/`dK`/`dV` (bwd) carry the previous tenant's values into the next op. The
same class of bug was already fixed once on the eager path (NVBug 6158764,
padding leak in the backward pass); the command-buffer path never got the
treatment.

Fix (`cudnn_thunk.cc`):
* Explicit path: on `RecordCreate`, emit one `CreateMemset` per output buffer and
  make the DNN graph command depend on them; on `RecordUpdate`, rebind those
  memset nodes to the new addresses via `UpdateMemset` before updating the graph
  command. The memset nodes are kept in `CommandState` keyed by the DNN graph
  command they precede, because the same thunk can be recorded into one command
  buffer more than once (unrolled loops), so a flat vector would alias.
* Traced fallback: zero the outputs inside the traced lambda, so the memsets are
  captured into the same nested command buffer as the graph launch.

Conservative choice: if the update path cannot find the memset nodes it recorded
(state evicted or never created), it returns `Internal(...)` rather than
silently skipping the zero-fill.

### 3. `force_deterministic` written after the instruction is fingerprinted

`xla/backends/gpu/transforms/cudnn_custom_call_compiler.cc:310-313` computed
`force_deterministic` from the module config and wrote it into the backend config
*inside* `BuildGraphForCustomCallToBackwardFMHA`. But that builder is invoked
from `CuDnnCustomCallVisitor::HandleCustomCall` **after**
`fingerprint_without_workspace` has already been computed (:552), and only on a
miss in `workspace_sizes_`.

The flag does change the built graph — `se::gpu::GetCudnnFlashAttentionBackwardOperationGraph`
forwards it into `EngineOptions{force_deterministic, ...}` and into the
deterministic-engine filtering at `xla/stream_executor/cuda/cuda_dnn.cc:4886`.

Failure scenario: compile a module with determinism required (e.g.
`--xla_gpu_deterministic_ops`) that contains two structurally identical backward
fMHA custom calls — i.e. any transformer with more than one identically-shaped
layer. The first call misses the workspace cache, builds the graph, gets
`force_deterministic: true` written into its backend config, and is registered in
`compilation_results_` under a fingerprint that includes the flag. The second call
hashes to the *same* `fingerprint_without_workspace` (it is computed before the
flag is written), takes the `// Cache hit.` branch at :576, and therefore never
gets the flag written at all. At emission time
`ThunkEmitter::EmitCuDnn` fingerprints that instruction *without* the flag
(`thunk_emitter.cc:1225`), and `CuDnnThunk::Initialize` then does
`params.src.dnn_compiled_graphs.at(fingerprint_)` (`cudnn_thunk.cc:82-83`) — an
`.at()` on a key that is not there.

Fix: move the write into a small `SetBackwardFMHADeterminism()` helper called from
`HandleCustomCall` before either fingerprint is computed; the graph builder now
just reads `config.force_deterministic()`. Both the workspace cache key and the
compilation-results key now cover the flag, and the flag is written on every
backward fMHA instruction regardless of cache hits.

## Evidence

The use-after-free is a plain lifetime bug visible in the source; no runtime repro
is needed and none is claimed. Note it is latent-but-real: the freed object's
storage is usually still warm, so it passes CI and fails under a different
allocator, sanitizer build or inlining decision.

No measured reproduction is claimed for any of the three fixes — all three were
established by reading the XLA and cuDNN sources.

## Testing

Added to `xla/backends/gpu/runtime/cudnn_thunk_test.cc`:

* `CuDnnThunkCmdBufTest.RecordCreateZeroFillsOutputExplicit`
* `CuDnnThunkCmdBufTest.RecordUpdateZeroFillsOutputExplicit`
* `CuDnnThunkCmdBufTest.RecordCreateZeroFillsOutputImplicit`

The existing fixture's output allocation now carries a `kPadElements` tail past
the matmul result that the DNN graph never writes — the stand-in for the padded
tokens of an fMHA output. `SetUp()` already poisons the whole output buffer with
`0xdeadbeef`, so after recording and submitting the command buffer the tail must
read back as zero. `FakeDnnGraph` gained a `write_output` flag so the traced path
can model a graph that writes nothing at all. All three tests fail before the
`cudnn_thunk.cc` change (tail stays `0xdeadbeef`) and pass after. The existing
four `Record()` tests are unaffected: they read back only the matmul region.

**These tests were written but not built or run** — this sandbox has no XLA build.
They are `xla_test(..., backends = ["gpu"], tags = ["cuda-only"])`, so they need a
GPU with cuDNN ≥ 9.7 and CUDA graph tracing support.

There is no test target for the `cudnn_custom_call_compiler` pass in the tree, so
fix 3 has no dedicated regression test. It is exercised end-to-end by
`xla/backends/gpu/tests/gpu_fused_mha_test.cc` when the module config requires
determinism.

Found by an integration audit of cuDNN SDPA consumers by the NVIDIA cuDNN team.

### CODE REVIEW

**Comment by YangXu1990uiuc** [comment]
Converting to draft: this needs to go through NVIDIA's internal review before upstreaming. I jumped that step — apologies for the noise. I'll mark it ready once internal review is done. The CLA is also outstanding on my side.

The technical content is unchanged; happy to take any early feedback in the meantime.
